File: ex8.py
#!/usr/bin/env python
import scipy.io
import numpy as np
import matplotlib.pyplot as plt
import estimategaussian as eg
import multivariategaussian as mvg
import visualizefit as vf
import selectthreshold as st

print('Visualizing example dataset for outlier detection.\n');

mat = scipy.io.loadmat('ex8data1.mat')
X = mat["X"]
Xval = mat["Xval"]
yval = mat["yval"].flatten()

#  Visualize the example dataset
plt.plot(X[:, 0], X[:, 1], 'bx', markersize=10, markeredgewidth=1)
plt.axis([0,30,0,30])
plt.xlabel('Latency (ms)')
plt.ylabel('Throughput (mb/s)')
plt.show(block=False)
input('Program paused. Press enter to continue.')

print('Visualizing Gaussian fit.\n')

mu, sigma2 = eg.estimategaussian(X)
p = mvg.multivariategaussian(X, mu, sigma2)

plt.close()
vf.visualizefit(X, mu, sigma2)
plt.xlabel('Latency (ms)')
plt.ylabel('Throughput (mb/s)')
plt.show(block=False)
input('Program paused. Press enter to continue.')

pval = mvg.multivariategaussian(Xval, mu, sigma2)

epsilon, F1 = st.selectthreshold(yval, pval)
print('Best epsilon found using cross-validation: {:e}'.format(epsilon))
print('Best F1 on Cross Validation Set:  {:f}'.format(F1))
print('   (you should see a value epsilon of about 8.99e-05)\n')
print('   (you should see a Best F1 value of  0.875000)\n\n');

# Find the outliers in the training set and plot the
outliers = p < epsilon

# interactive graphs
plt.ion()

#  Draw a red circle around those outliers

# plt.scatter(X[outliers, 0], X[outliers, 1], s=325, facecolors='none', edgecolors='r')
plt.plot(X[outliers, 0], X[outliers, 1], 'ro', linewidth=2, markersize=18, fillstyle='none', markeredgewidth=1)
plt.show(block=False)

input('Program paused. Press enter to continue.')

mat = scipy.io.loadmat('ex8data2.mat')
X = mat["X"]
Xval = mat["Xval"]
yval = mat["yval"].flatten()

#  Apply the same steps to the larger dataset
mu, sigma2 = eg.estimategaussian(X)

#  Training set
p = mvg.multivariategaussian(X, mu, sigma2)

#  Cross-validation set
pval = mvg.multivariategaussian(Xval, mu, sigma2)

#  Find the best threshold
epsilon, F1 = st.selectthreshold(yval, pval)

print('Best epsilon found using cross-validation: {:e}'.format(epsilon))
print('Best F1 on Cross Validation Set:  {:f}'.format(F1))
print('   (you should see a value epsilon of about 1.38e-18)\n')
print('   (you should see a value epsilon of about 1.38e-18)\n');
print('   (you should see a Best F1 value of 0.615385)\n');
print('# Outliers found: {:d}'.format(np.sum((p < epsilon).astype(int))))

File: ex8.R
rm(list=ls())
sources <- c("bsxfun.R","estimateGaussian.R","meshgrid.R",
             "multivariateGaussian.R","pinv.R","selectThreshold.R",
             "visualizeFit.R")

for (i in 1:length(sources)) {
  cat(paste("Loading ",sources[i],"\n"))
  source(sources[i])
}

## ------------------ Part 1: Load Example Dataset  --------------------

cat(sprintf('Visualizing example dataset for outlier detection.\n\n'))

#  The following command loads the dataset. 
load("ex8data1.Rda")
list2env(data,.GlobalEnv)
rm(data)

#  Visualize the example dataset
plot(X[,1], X[,2], xlim=c(0,30),ylim=c(0,30), pch=4, col="blue",
     xlab='Latency (ms)', ylab='Throughput (mb/s)')

cat(sprintf('Program paused. Press enter to continue.\n'))
line <- readLines(con = stdin(),1)


## ------------------ Part 2: Estimate the dataset statistics --------------------

cat(sprintf('Visualizing Gaussian fit.\n\n'))

#  Estimate mu and sigma2
eG <- estimateGaussian(X)
mu <- eG$mu
sigma2 <- eG$sigma2

#  Returns the density of the multivariate normal at each data point (row) 
#  of X
p <- multivariateGaussian(X, mu, sigma2)

#  Visualize the fit
visualizeFit(X,  mu, sigma2)

cat(sprintf('Program paused. Press enter to continue.\n'))
line <- readLines(con = stdin(),1)

## ------------------ Part 3: Find Outliers --------------------

pval <- multivariateGaussian(Xval, mu, sigma2)

sT <- selectThreshold(yval, pval)
epsilon <- sT$bestEpsilon
F1 <- sT$bestF1

cat(sprintf('Best epsilon found using cross-validation: %e\n', epsilon))
cat(sprintf('Best F1 on Cross Validation Set:  %f\n', F1))
cat(sprintf('(you should see a value epsilon of about 8.99e-05)\n\n'))

#  Find the outliers in the training set and plot the
outliers <- which(p < epsilon)

#  Draw a red circle around those outliers
points(X[outliers, 1], X[outliers, 2], col="red", lwd=2, cex=2 )

cat(sprintf('Program paused. Press enter to continue.\n'))
line <- readLines(con = stdin(),1)

## ------------------ Part 4: Multidimensional Outliers --------------------

load("ex8data2.Rda")
list2env(data,.GlobalEnv)
rm(data)

#  Apply the same steps to the larger dataset
eG <- estimateGaussian(X)
mu <- eG$mu
sigma2 <- eG$sigma2
rm(eG)

#  Training set 
p <- multivariateGaussian(X, mu, sigma2)

#  Cross-validation set
pval <- multivariateGaussian(Xval, mu, sigma2)

#  Find the best threshold
sT <- selectThreshold(yval, pval)
epsilon <- sT$bestEpsilon
F1 <- sT$bestF1

cat(sprintf('Best epsilon found using cross-validation: %e\n', epsilon))
cat(sprintf('Best F1 on Cross Validation Set:  %f\n', F1))
cat(sprintf('# Outliers found: %d\n', sum(p < epsilon)))
cat(sprintf('   (you should see a value epsilon of about 1.38e-18)\n\n'))
line <- readLines(con = stdin(),1)